drud/ddev
# Don't restart the ddev-router all the time, fixes #1236

## The Problem/Issue/Bug:

OP #1236 - We stop the router and check for potential port conflicts with every single ddev start, and it takes a few seconds in the start. It's unnecessary.

## How this PR Solves The Problem:

* Calculate required ports to check for conflicts and only check new ports.
* Just let `docker-compose up` do the job of deciding whether to recreate ddev-router.

## Manual Testing Instructions:

Use a few projects with different ports.
Use `ddev start` on them and monitor ddev-router behavior. 

You can force the occupy-port test with the nc tool, like `nc -l 8080` to force occupation of port 8080 to see if the port check does the right thing on ddev start.


## Automated Testing Overview:

TestGetExposedContainerPorts() was added for the new GetExposedContainerPorts() utility.
Our existing tests should do the job of testing port overrides and such, as the behavior mostly stays on parallel.

## Related Issue Link(s):

OP #1236
